src/state.rs

```diff
@@ -341,9 +387,51 @@ impl DemandProxyState {
                     .on_demand_dns_cache_misses
                     .get_or_create(&labels)
                     .inc();
-                // TODO: optimize so that if multiple requests to the same hostname come in at the same time,
+                // optimize so that if multiple requests to the same hostname come in at the same time,
                 // we don't start more than one background on-demand DNS task
-                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                if state
```

Comment:
I think we need to simply all of this ... it's getting pretty complex. It might be cleaner to separate out the DNS resolution logic into a separate, dedicated structure. For example:

```rust
struct DnsResolver {
  state: Arc<RwLock<ProxyState>>,
  notifyMap: Mutex<HashMap<string,Arc<Notify>>
}

impl DnsResolver {
  async fn resolve_host(&self, host) Option<ResolvedDns>{
    // If we don't have the addresses (i.e. get_ips_for_hostname)
    // then:
    //     Get or create a `Notify` for the host in `notifyMap`.
    //     If created:
    //         resolve DNS and notify waiters
    //     else:
    //         wait for notify
    // Return the addresses (i.e. get_ips_for_hostname)
  }
}

```

Then the code for `loadbalance_for_hostname` would be reduced to something like:


```rust
let rdns = match self.dnsResolver.resolve_host(&workload.hostname)
...
```

Reply:
Hi @nmittler, thanks for your suggestion and I also followed your idea to change the code to make it simple. However, I find that the function `resolve_on_demand_dns` which really do the DNS resolving task must be done inside the `DemandProxyState` because it refers to many data owned by `DemandProxyState`, such as insert the resolved ip addresses into HashMap of `by_hostname` which is belonged to `DemandProxyState`.

Any idea?

Reply:
Can you just pass in a function reference? Basically, `DnsResolver` just needs some function that actually resolves DNS.

Reply:
I'm open to other opinions on this @howardjohn @hzxuzhonghu

Reply:
> Can you just pass in a function reference? Basically, `DnsResolver` just needs some function that actually resolves DNS.

Okay, l will try it out, I understand your mean is that move all checking logic into `DnsResolver`, then call `resolve_on_demand_dns` of `DemandProxyState`,  am I right?

Reply:
> Can you just pass in a function reference? Basically, `DnsResolver` just needs some function that actually resolves DNS.

Okay, @nmittler, l will try it out, I understand your mean is that move all checking logic into `DnsResolver`, then call `resolve_on_demand_dns` of `DemandProxyState`,  am I right?

Reply:
> Can you just pass in a function reference? Basically, `DnsResolver` just needs some function that actually resolves DNS.

Okay, @nmittler, I understand your mean is that move all checking logic into `DnsResolver`, then call `resolve_on_demand_dns` of `DemandProxyState`,  am I right?

Reply:
> I think we need to simply all of this ... it's getting pretty complex. It might be cleaner to separate out the DNS resolution logic into a separate, dedicated structure. For example:
> 
> ```rust
> struct DnsResolver {
>   state: Arc<RwLock<ProxyState>>,
>   notifyMap: Mutex<HashMap<string,Arc<Notify>>
> }
> 
> impl DnsResolver {
>   async fn resolve_host(&self, host) Option<ResolvedDns>{
>     // If we don't have the addresses (i.e. get_ips_for_hostname)
>     // then:
>     //     Get or create a `Notify` for the host in `notifyMap`.
>     //     If created:
>     //         resolve DNS and notify waiters
>     //     else:
>     //         wait for notify
>     // Return the addresses (i.e. get_ips_for_hostname)
>   }
> }
> ```
> 
> Then the code for `loadbalance_for_hostname` would be reduced to something like:
> 
> ```rust
> let rdns = match self.dnsResolver.resolve_host(&workload.hostname)
> ...
> ```

Hi @nmittler, I added more description in `load_balance_for_hostname`, so do you mean I need to move the logic of `Err` branch into another function?

Reply:
Hi @nmittler, I made a big change based on your command, all DNS resolving relates are now in a struct `DnsResolver`. Please help to review again, thanks Mittler!

Reply:
Hi @nmittler, I made a change based on your command, all DNS resolving relates are now in a struct `DnsResolver`. Please help to review again, thanks Mittler!